crates/emitter/src/reference_op_emitter.rs

```diff
@@ -0,0 +1,452 @@
+        Self { kind }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+enum CallKind {
+    Normal,
+    // FIXME: Support eval, Function#call, Function#apply etc.
+}
+
+// See *ReferenceEmitter.
+// This uses struct to hide the details from the consumer.
+#[derive(Debug)]
+#[must_use]
+pub struct CallReference {
+    kind: CallKind,
+}
+impl CallReference {
+    fn new(kind: CallKind) -> Self {
+        Self { kind }
+    }
+}
+
+// Struct for emitting bytecode for get `name` operation.
+pub struct GetNameEmitter<'alloc> {
+    pub name: &'alloc str,
+}
+impl<'alloc> GetNameEmitter<'alloc> {
+    pub fn emit(self, emitter: &mut AstEmitter) {
+        let name_index = emitter.emit.get_atom_index(self.name);
+
+        //              [stack]
+
+        // FIXME: Support non-global case.
+        emitter.emit.get_g_name(name_index);
+        //              [stack] VAL
+    }
+}
+
+// Struct for emitting bytecode for get `obj.key` operation.
+pub struct GetPropEmitter<'alloc, F>
+where
+    F: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+{
+    pub obj: F,
+    pub key: &'alloc str,
+}
+impl<'alloc, F> GetPropEmitter<'alloc, F>
+where
+    F: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+{
+    pub fn emit(self, emitter: &mut AstEmitter) -> Result<(), EmitError> {
+        let key_index = emitter.emit.get_atom_index(self.key);
+
+        //              [stack]
+
+        (self.obj)(emitter)?;
+        //              [stack] OBJ
+
+        emitter.emit.get_prop(key_index);
+        //              [stack] VAL
+
+        Ok(())
+    }
+}
+
+// Struct for emitting bytecode for get `super.key` operation.
+pub struct GetSuperPropEmitter<'alloc, F>
+where
+    F: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+{
+    pub this: F,
+    pub key: &'alloc str,
+}
+impl<'alloc, F> GetSuperPropEmitter<'alloc, F>
+where
+    F: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+{
+    pub fn emit(self, emitter: &mut AstEmitter) -> Result<(), EmitError> {
+        let key_index = emitter.emit.get_atom_index(self.key);
+
+        //              [stack]
+
+        (self.this)(emitter)?;
+        //              [stack] THIS
+
+        emitter.emit.callee();
+        //              [stack] THIS CALLEE
+
+        emitter.emit.super_base();
+        //              [stack] THIS OBJ
+
+        emitter.emit.get_prop_super(key_index);
+        //              [stack] VAL
+
+        Ok(())
+    }
+}
+
+// Struct for emitting bytecode for get `obj[key]` operation.
+pub struct GetElemEmitter<F1, F2>
+where
+    F1: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+    F2: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+{
+    pub obj: F1,
+    pub key: F2,
+}
+impl<F1, F2> GetElemEmitter<F1, F2>
+where
+    F1: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+    F2: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+{
+    pub fn emit(self, emitter: &mut AstEmitter) -> Result<(), EmitError> {
+        //              [stack]
+
+        (self.obj)(emitter)?;
```

Comment:
It would be nice to use `debug_assert_eq!` to check that each of these lambdas bumps the stack depth by 1. Optional.